File: src/commands/prune.ts
import { existsSync } from "node:fs";
import { join } from "node:path";
import { formatAllocationEntry } from "../format";
import type { Allocation } from "../state/state";
import { loadState, withState } from "../state/state";

export interface PruneOptions {
  dryRun: boolean;
}

interface OrphanInfo {
  allocation: Allocation;
  reason: string;
}

function classifyOrphans(allocations: Allocation[]): OrphanInfo[] {
  const orphans: OrphanInfo[] = [];
  for (const a of allocations) {
    if (!existsSync(a.dir)) {
      orphans.push({ allocation: a, reason: "directory missing" });
      continue;
    }
    if (!existsSync(join(a.dir, "port-pool.config.json"))) {
      orphans.push({ allocation: a, reason: "no port-pool.config.json" });
    }
  }
  return orphans;
}

function reportOrphans(orphans: OrphanInfo[], verb: string): void {
  if (orphans.length === 0) {
    console.log("No orphaned allocations");
    return;
  }
  console.log(`${verb} ${orphans.length} allocation(s):`);
  for (const o of orphans) {
    console.log(`  - ${o.allocation.dir} (${o.reason})`);
    for (const entry of o.allocation.entries) {
      console.log(`    ${formatAllocationEntry(entry)}`);
    }
  }
}

export function prune(opts: PruneOptions): void {
  if (opts.dryRun) {
    reportOrphans(classifyOrphans(loadState().allocations), "Would prune");
    return;
  }

  withState((state) => {
    const orphans = classifyOrphans(state.allocations);
    if (orphans.length > 0) {
      const orphanDirs = new Set(orphans.map((o) => o.allocation.dir));
      state.allocations = state.allocations.filter((a) => !orphanDirs.has(a.dir));
    }
    reportOrphans(orphans, "Pruned");
  });
}

File: internal/cmd/prune.go
package cmd

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	"github.com/sylophi/port-pool/internal/state"
)

const (
	pruneUsage       = "usage: port-pool prune [--dry-run]"
	reasonDirMissing = "directory missing"
	reasonNoConfig   = "no port-pool.config.json"
)

type orphanInfo struct {
	allocation state.Allocation
	reason     string
}

func Prune(args []string) error {
	args, dryRun := extractBoolFlag(args, "dry-run")
	for _, a := range args {
		if strings.HasPrefix(a, "--") {
			return fmt.Errorf("unknown flag: %s\n%s", a, pruneUsage)
		}
	}
	if len(args) > 0 {
		return fmt.Errorf("unexpected arguments: %v\n%s", args, pruneUsage)
	}

	if dryRun {
		s, err := state.Load()
		if err != nil {
			return err
		}
		reportOrphans(classifyOrphans(s.Allocations), "Would prune")
		return nil
	}

	return state.WithState(func(s *state.State) error {
		orphans := classifyOrphans(s.Allocations)
		if len(orphans) > 0 {
			orphanDirs := make(map[string]struct{}, len(orphans))
			for _, o := range orphans {
				orphanDirs[o.allocation.Dir] = struct{}{}
			}
			kept := s.Allocations[:0]
			for _, a := range s.Allocations {
				if _, dropped := orphanDirs[a.Dir]; !dropped {
					kept = append(kept, a)
				}
			}
			s.Allocations = kept
		}
		reportOrphans(orphans, "Pruned")
		return nil
	})
}

func classifyOrphans(allocs []state.Allocation) []orphanInfo {
	var orphans []orphanInfo
	for _, a := range allocs {
		if _, err := os.Stat(a.Dir); errors.Is(err, fs.ErrNotExist) {
			orphans = append(orphans, orphanInfo{allocation: a, reason: reasonDirMissing})
			continue
		}
		cfgPath := filepath.Join(a.Dir, "port-pool.config.json")
		if _, err := os.Stat(cfgPath); errors.Is(err, fs.ErrNotExist) {
			orphans = append(orphans, orphanInfo{allocation: a, reason: reasonNoConfig})
		}
	}
	return orphans
}

func reportOrphans(orphans []orphanInfo, verb string) {
	if len(orphans) == 0 {
		fmt.Println("No orphaned allocations")
		return
	}
	fmt.Printf("%s %d allocation(s):\n", verb, len(orphans))
	for _, o := range orphans {
		fmt.Printf("  - %s (%s)\n", o.allocation.Dir, o.reason)
		fmt.Printf("    %s\n", o.allocation.PortString())
	}
}
